lnrpc/routerrpc/router_server.go

```diff
@@ -266,3 +269,169 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+		}, nil
+	}
+
+	// In the failure case, marshall the failure message to the rpc format
+	// before returning it to the caller.
+	rpcErr, err := marshallError(err)
+	if err != nil {
+		return nil, err
+	}
+
+	return &SendToRouteResponse{
+		Failure: rpcErr,
+	}, nil
+}
+
+// marshallError marshall an error as received from the switch to rpc structs
+// suitable for returning to the caller of an rpc method.
+//
+// Because of difficulties with using protobuf oneof constructs in some
+// languages, the decision was made here to use a single message format for all
+// failure messages with some fields left empty depending on the failure type.
+func marshallError(sendError error) (*Failure, error) {
+	response := &Failure{}
+
+	fErr, ok := sendError.(*htlcswitch.ForwardingError)
+	if !ok {
+		return nil, sendError
+	}
+
+	switch onionErr := fErr.FailureMessage.(type) {
+
+	case *lnwire.FailUnknownPaymentHash:
+		response.Code = Failure_UNKNOWN_PAYMENT_HASH
+
+	case *lnwire.FailIncorrectPaymentAmount:
+		response.Code = Failure_INCORRECT_PAYMENT_AMOUNT
+
+	case *lnwire.FailFinalIncorrectCltvExpiry:
+		response.Code = Failure_FINAL_INCORRECT_CLTV_EXPIRY
+		response.CltvExpiry = onionErr.CltvExpiry
+
+	case *lnwire.FailFinalIncorrectHtlcAmount:
+		response.Code = Failure_FINAL_INCORRECT_HTLC_AMOUNT
+		response.HtlcMsat = uint64(onionErr.IncomingHTLCAmount)
+
+	case *lnwire.FailFinalExpiryTooSoon:
+		response.Code = Failure_FINAL_EXPIRY_TOO_SOON
+
+	case *lnwire.FailInvalidRealm:
+		response.Code = Failure_INVALID_REALM
+
+	case *lnwire.FailExpiryTooSoon:
+		response.Code = Failure_EXPIRY_TOO_SOON
+		response.ChannelUpdate = marshallChannelUpdate(&onionErr.Update)
+
+	case *lnwire.FailInvalidOnionVersion:
+		response.Code = Failure_INVALID_ONION_VERSION
+		response.OnionSha_256 = onionErr.OnionSHA256[:]
+
+	case *lnwire.FailInvalidOnionHmac:
+		response.Code = Failure_INVALID_ONION_HMAC
+		response.OnionSha_256 = onionErr.OnionSHA256[:]
+
+	case *lnwire.FailInvalidOnionKey:
+		response.Code = Failure_INVALID_ONION_KEY
+		response.OnionSha_256 = onionErr.OnionSHA256[:]
+
+	case *lnwire.FailAmountBelowMinimum:
+		response.Code = Failure_AMOUNT_BELOW_MINIMUM
+		response.ChannelUpdate = marshallChannelUpdate(&onionErr.Update)
+		response.HtlcMsat = uint64(onionErr.HtlcMsat)
+
+	case *lnwire.FailFeeInsufficient:
+		response.Code = Failure_FEE_INSUFFICIENT
+		response.ChannelUpdate = marshallChannelUpdate(&onionErr.Update)
+		response.HtlcMsat = uint64(onionErr.HtlcMsat)
+
+	case *lnwire.FailIncorrectCltvExpiry:
+		response.Code = Failure_INCORRECT_CLTV_EXPIRY
+		response.ChannelUpdate = marshallChannelUpdate(&onionErr.Update)
+		response.CltvExpiry = onionErr.CltvExpiry
+
+	case *lnwire.FailChannelDisabled:
+		response.Code = Failure_CHANNEL_DISABLED
+		response.ChannelUpdate = marshallChannelUpdate(&onionErr.Update)
+
+		// todo: is this the same flag as in channel update?
+		// is it every filled properly?
+		response.Flags = uint32(onionErr.Flags)
```

Comment:
Atm this is the raw set of flags as would be populated in a channel update message.

Reply:
Why is it a separate field when the `Update` field already contains a channel update including the flags?

Reply:
Historically I think it's due to then being developed more or less independently. However, today there's no a split in that only have of what's now the flags field in the channel update signals the directionality. The other half is now used to signal the presence of certain optional fields.

Reply:
I also see `lnd` never sends this failure. 

Removed the todo and just pass back as is.